src/middleware/query-advanced.js

```diff
@@ -0,0 +1,260 @@
+import Backoff from 'backo';
+import invariant from 'invariant';
+import get from 'lodash.get';
+import identity from 'lodash.identity';
+import includes from 'lodash.includes';
+import pickBy from 'lodash.pickby';
+
+import {
+    requestStart,
+    requestFailure,
+    requestSuccess,
+    mutateStart,
+    mutateFailure,
+    mutateSuccess,
+} from '../actions';
+import * as actionTypes from '../constants/action-types';
+import * as httpMethods from '../constants/http-methods';
+import * as statusCodes from '../constants/status-codes';
+import { reconcileQueryKey } from '../lib/get-query-key';
+
+const updateEntities = (update, entities, transformed) => {
+    // If update, not supplied, then no change to entities should be made
+
+    return Object.keys(update || {}).reduce((accum, key) => {
+        accum[key] = update[key]((entities || {})[key], (transformed || {})[key]);
+
+        return accum;
+    }, {});
+};
+
+const optimisticUpdateEntities = (optimisticUpdate, entities) => {
+    return Object.keys(optimisticUpdate).reduce((accum, key) => {
+        if (optimisticUpdate[key]) {
+            accum[key] = optimisticUpdate[key](entities[key]);
+        } else {
+            accum[key] = entities[key];
+        }
+
+        return accum;
+    }, {});
+};
+
+const defaultConfig = {
+    backoff: {
+        maxAttempts: 5,
+        minDuration: 300,
+        maxDuration: 5000,
+    },
+    retryableStatusCodes: [
+        statusCodes.UNKNOWN, // normally means a failed connection
+        statusCodes.REQUEST_TIMEOUT,
+        statusCodes.TOO_MANY_REQUESTS, // hopefully backoff stops this getting worse
+        statusCodes.SERVICE_UNAVAILABLE,
+        statusCodes.GATEWAY_TIMEOUT,
+    ],
+};
+
+const getPendingQueries = (queries) => {
+    return pickBy(queries, (query) => query.isPending);
+};
+
+const resOk = (status) => Math.floor(status / 100) === 2;
+
+const queryMiddlewareAdvanced = (queriesSelector, entitiesSelector, networkAdapter, config = defaultConfig) => {
```

Comment:
If we pull `networkAdapter` out like so:

`const queryMiddlewareAdvanced = (networkAdapter) => (queriesSelector, entitiesSelector, config = defaultConfig)`

Then the default middleware will be much simpler:

`const queryMiddleware = queryAdvanced(superagentAdapter);`